File: scoring/engine.py
"""
scoring/engine.py — מנוע הניקוד.

חוזה:
    ScoringEngine(ruleset).score(lineup, performances, outcomes) -> LineupScore

טהור לחלוטין. אותם קלטים -> אותה תוצאה, תמיד. אפשר להריץ אותו על
מחזור מלפני שנה ולקבל בדיוק את אותו מספר.

שלושת רכיבי הניקוד:
    1. ציון אישי   — סכימת פעולות של כל שחקן בהרכב
    2. בונוס תוצאה — ניצחון של הקבוצה האמיתית (+4) / תיקו (+1)
    3. שער וירטואלי — כל 2 שערים קבוצתיים במציאות = +5
"""
from __future__ import annotations

import logging
from collections import defaultdict
from typing import Iterable, Mapping, Sequence

from .models import (
    Lineup, LineupScore, LineupSlot, PlayerPerformance, PlayerScore,
    Reason, ScoreLine, TeamOutcome,
)
from .rules import RuleSet
from .validation import assert_valid

log = logging.getLogger("scoring")


class ScoringEngine:
    """חסר-מצב (stateless) מעבר ל-ruleset. בטוח לשימוש חוזר ומקבילי."""

    def __init__(self, ruleset: RuleSet):
        self.rules = ruleset

    # ==================================================================
    # API ציבורי
    # ==================================================================
    def score(
        self,
        lineup: Lineup,
        performances: Mapping[str, PlayerPerformance],
        outcomes: Mapping[str, TeamOutcome],
        *,
        validate: bool = True,
    ) -> LineupScore:
        """
        performances : player_id -> ביצועים במחזור
        outcomes     : team_id   -> תוצאת הקבוצה במחזור
        """
        if validate:
            assert_valid(lineup, self.rules)

        result = LineupScore(
            lineup_id=lineup.lineup_id,
            gameweek_id=lineup.gameweek_id,
            ruleset_version=self.rules.version,
        )

        for slot in lineup.starters:
            perf = performances.get(slot.player_id)
            outcome = outcomes.get(slot.team_id)
            result.players.append(self._score_player(slot, perf, outcome))

        result.bonus_lines.extend(self._virtual_goals(lineup.starters, outcomes))
        return result

    # ==================================================================
    # 1 + 2. ציון אישי ובונוס תוצאה (נצברים על השחקן)
    # ==================================================================
    def _score_player(self, slot: LineupSlot,
                      perf: PlayerPerformance | None,
                      outcome: TeamOutcome | None) -> PlayerScore:
        ps = PlayerScore(player_id=slot.player_id, team_id=slot.team_id,
                         position=slot.position, is_captain=slot.is_captain)
        p = self.rules.personal
        pos = slot.position

        if perf is None or not perf.played:
            # שחקן שלא שיחק: 0 נקודות אישיות. בונוס התוצאה נשלט ב-require_minutes.
            log.debug("אין ביצועים לשחקן %s", slot.player_id)
        else:
            add = ps.lines.append

            # דקות משחק
            if perf.minutes > 0:
                pts = p.minutes_60_plus if perf.minutes >= 60 else p.minutes_played
                add(ScoreLine(Reason.MINUTES, pts, count=perf.minutes,
                              player_id=slot.player_id))

            # שערים — משוקללים לפי עמדה
            if perf.goals:
                unit = float(p.goal.get(pos, 0))
                add(ScoreLine(Reason.GOAL, unit * perf.goals, count=perf.goals,
                              player_id=slot.player_id, meta={"per_goal": unit}))

            if perf.assists:
                add(ScoreLine(Reason.ASSIST, p.assist * perf.assists,
                              count=perf.assists, player_id=slot.player_id))

            # קלין-שיט — רק למי ששיחק מספיק דקות
            if perf.clean_sheet and perf.minutes >= p.clean_sheet_min_minutes:
                unit = float(p.clean_sheet.get(pos, 0))
                if unit:
                    add(ScoreLine(Reason.CLEAN_SHEET, unit,
                                  player_id=slot.player_id))

            # ספיגות (שוערים ומגנים)
            if perf.goals_conceded >= p.goals_conceded_per:
                unit = float(p.goals_conceded.get(pos, 0))
                chunks = perf.goals_conceded // p.goals_conceded_per
                if unit:
                    add(ScoreLine(Reason.GOALS_CONCEDED, unit * chunks,
                                  count=perf.goals_conceded,
                                  player_id=slot.player_id))

            # הצלות
            if perf.saves >= p.saves_per:
                chunks = perf.saves // p.saves_per
                add(ScoreLine(Reason.SAVES, p.saves_points * chunks,
                              count=perf.saves, player_id=slot.player_id))

            if perf.penalties_saved:
                add(ScoreLine(Reason.PENALTY_SAVED,
                              p.penalty_saved * perf.penalties_saved,
                              count=perf.penalties_saved, player_id=slot.player_id))
            if perf.penalties_missed:
                add(ScoreLine(Reason.PENALTY_MISSED,
                              p.penalty_missed * perf.penalties_missed,
                              count=perf.penalties_missed, player_id=slot.player_id))

            # עונשין
            if perf.own_goals:
                add(ScoreLine(Reason.OWN_GOAL, p.own_goal * perf.own_goals,
                              count=perf.own_goals, player_id=slot.player_id))
            if perf.yellow_cards:
                add(ScoreLine(Reason.YELLOW, p.yellow_card * perf.yellow_cards,
                              count=perf.yellow_cards, player_id=slot.player_id))
            if perf.red_cards:
                add(ScoreLine(Reason.RED, p.red_card * perf.red_cards,
                              count=perf.red_cards, player_id=slot.player_id))

        # --- בונוס תוצאה של הקבוצה האמיתית ---
        bonus = self._result_bonus(perf, outcome)
        if bonus is not None:
            ps.lines.append(bonus)

        # --- קפטן: מכפיל על כל מה שהצטבר לשחקן הזה ---
        if slot.is_captain and self.rules.captain_multiplier != 1:
            base = ps.subtotal
            extra = base * (self.rules.captain_multiplier - 1)
            if extra:
                ps.lines.append(ScoreLine(
                    Reason.CAPTAIN, round(extra, 2), player_id=slot.player_id,
                    meta={"multiplier": self.rules.captain_multiplier, "base": base}))

        return ps

    def _result_bonus(self, perf: PlayerPerformance | None,
                      outcome: TeamOutcome | None) -> ScoreLine | None:
        if outcome is None:
            return None
        rb = self.rules.result_bonus
        if rb.require_minutes and (perf is None or perf.minutes < rb.require_minutes):
            return None
        points = {"W": rb.win, "D": rb.draw, "L": rb.loss}[outcome.result]
        if points == 0:
            return None
        return ScoreLine(Reason.RESULT_BONUS, float(points),
                         team_id=outcome.team_id, meta={"result": outcome.result})

    # ==================================================================
    # 3. שערי הרכב וירטואליים
    # ==================================================================
    def _virtual_goals(self, slots: Sequence[LineupSlot],
                       outcomes: Mapping[str, TeamOutcome]) -> list[ScoreLine]:
        """
        כל N שערים שכבשו הקבוצות שבהרכב = שער וירטואלי.

        pooled   : סוכמים הכל, מחלקים פעם אחת. שאריות נשמרות בין קבוצות.
        per_team : כל קבוצה לחוד, שאריות נזרקות.
        """
        v = self.rules.virtual_goal
        if v.goals_per_virtual <= 0:
            return []

        # קבוצות ייחודיות בלבד — גם אם החוקים יאפשרו בעתיד 2 מאותה קבוצה,
        # שערי הקבוצה לא ייספרו פעמיים.
        team_goals: dict[str, int] = {}
        for slot in slots:
            outcome = outcomes.get(slot.team_id)
            if outcome is not None:
                team_goals[slot.team_id] = outcome.goals_for

        if not team_goals:
            return []

        if v.aggregation == "per_team":
            lines: list[ScoreLine] = []
            for team_id, goals in sorted(team_goals.items()):
                count = goals // v.goals_per_virtual
                if count:
                    lines.append(ScoreLine(
                        Reason.VIRTUAL_GOAL, v.points * count, count=count,
                        team_id=team_id, meta={"team_goals": goals}))
            return lines

        total_goals = sum(team_goals.values())
        count = total_goals // v.goals_per_virtual
        if not count:
            return []
        return [ScoreLine(
            Reason.VIRTUAL_GOAL, v.points * count, count=count,
            meta={"team_goals_total": total_goals,
                  "remainder": total_goals % v.goals_per_virtual,
                  "per_team": team_goals})]


# ======================================================================
# מצרפים: טבלת המחזור
# ======================================================================
def rank_gameweek(scores: Iterable[LineupScore]) -> list[tuple[int, LineupScore]]:
    """דירוג עם טיפול בשוויון (אותו ניקוד = אותו מקום)."""
    ordered = sorted(scores, key=lambda s: s.total_points, reverse=True)
    ranked: list[tuple[int, LineupScore]] = []
    last_points, last_rank = None, 0
    for idx, score in enumerate(ordered, start=1):
        if score.total_points != last_points:
            last_rank, last_points = idx, score.total_points
        ranked.append((last_rank, score))
    return ranked


# ======================================================================
# גשר ל-DB (השכבה היחידה שיודעת על SQL)
# ======================================================================
def build_inputs_from_rows(
    perf_rows: Iterable[Mapping],
    match_rows: Iterable[Mapping],
) -> tuple[dict[str, PlayerPerformance], dict[str, TeamOutcome]]:
    """
    ממיר תוצאות שאילתה למבני הקלט של המנוע.
    perf_rows  : core.player_match_stats מסונן למחזור
    match_rows : core.v_team_match_results מסונן למחזור
    שחקן ששיחק שני משחקים באותו מחזור (דחייה שהושלמה) — נצבר.
    """
    perfs: dict[str, PlayerPerformance] = {}
    acc: dict[str, dict] = defaultdict(lambda: defaultdict(int))

    for r in perf_rows:
        a = acc[r["player_id"]]
        a["team_id"] = r["team_id"]
        a["position"] = r["position"]
        for k in ("minutes", "goals", "assists", "own_goals", "yellow_cards",
                  "red_cards", "saves", "penalties_saved", "penalties_missed",
                  "goals_conceded"):
            a[k] += r.get(k, 0) or 0
        a["clean_sheets"] += 1 if r.get("clean_sheet") else 0

    for player_id, a in acc.items():
        perfs[player_id] = PlayerPerformance(
            player_id=player_id, team_id=a["team_id"], position=a["position"],
            minutes=a["minutes"], goals=a["goals"], assists=a["assists"],
            own_goals=a["own_goals"], yellow_cards=a["yellow_cards"],
            red_cards=a["red_cards"], saves=a["saves"],
            penalties_saved=a["penalties_saved"],
            penalties_missed=a["penalties_missed"],
            goals_conceded=a["goals_conceded"],
            clean_sheet=a["clean_sheets"] > 0 and a["goals_conceded"] == 0,
            played=a["minutes"] > 0,
        )

    outcomes: dict[str, TeamOutcome] = {}
    for r in match_rows:
        team_id = r["team_id"]
        prev = outcomes.get(team_id)
        # מחזור עם שני משחקים לאותה קבוצה: צוברים שערים, התוצאה הטובה קובעת
        if prev is None:
            outcomes[team_id] = TeamOutcome(team_id, r["result"],
                                            r["goals_for"], r["goals_against"])
        else:
            best = max(prev.result, r["result"], key=lambda x: {"W": 2, "D": 1, "L": 0}[x])
            outcomes[team_id] = TeamOutcome(
                team_id, best,
                prev.goals_for + r["goals_for"],
                prev.goals_against + r["goals_against"])
    return perfs, outcomes

File: supabase/functions/dubid-score-gameweek/_lib/scoring/engine.ts
/**
 * ⚠ נוצר אוטומטית — אל תערכו כאן.
 *
 * הועתק מ-src/lib/ על ידי scripts/sync-edge-shared.mjs.
 * כל שינוי כאן יימחק בפריסה הבאה. ערכו את המקור.
 */
/**
 * scoring/engine.ts — מנוע הניקוד של דוביד.
 *
 *   scoreLineup(lineup, performances, outcomes, rules) -> LineupScore
 *
 * טהור לחלוטין: אין DB, אין fetch, אין Date.now(). אותם קלטים מחזירים
 * תמיד אותו מספר — אפשר לחשב מחדש מחזור מלפני שנה ולקבל בדיוק אותו ציון.
 * זה מה שמאפשר להריץ אותו גם בדפדפן (תצוגה חיה) וגם ב-Edge Function
 * (החישוב הרשמי) בלי סיכון לפער בין המספרים.
 *
 * ארבעת רכיבי הניקוד:
 *   1. ציון אישי       — פעולות השחקן, משוקללות לפי עמדה
 *   2. בונוס תוצאה     — ניצחון הקבוצה האמיתית +4 / תיקו +1
 *   3. שער וירטואלי    — כל 2 שערים של קבוצות ההרכב = +5
 *   4. הקפטן הדובידי   — ×3, חסינות כרטיסים, והעברה לסגן
 */
import {
  Reason,
  type Lineup,
  type LineupScore,
  type LineupSlot,
  type PlayerPerformance,
  type PlayerScore,
  type ScoreLine,
  type TeamOutcome,
} from './types.ts';
import { IL_PREMIER, type RuleSet } from './rules.ts';
import { assertValidLineup, starters } from './validate.ts';

const round2 = (n: number) => Math.round(n * 100) / 100;
const sum = (lines: ScoreLine[]) => lines.reduce((a, l) => a + l.points, 0);

export interface ScoreOptions {
  /** ברירת מחדל true. כבו רק לחישובי "מה היה קורה אם". */
  validate?: boolean;
  locale?: 'he' | 'en';
}

export function scoreLineup(
  lineup: Lineup,
  performances: Map<string, PlayerPerformance> | Record<string, PlayerPerformance>,
  outcomes: Map<string, TeamOutcome> | Record<string, TeamOutcome>,
  rules: RuleSet = IL_PREMIER,
  options: ScoreOptions = {},
): LineupScore {
  const { validate = true, locale = 'he' } = options;

  // ★ שכבת ההגנה: אין ניקוד להרכב לא חוקי. גם לא "בערך".
  if (validate) assertValidLineup(lineup, rules, locale);

  const perf = toMap(performances);
  const out = toMap(outcomes);
  const xi = starters(lineup);

  // --- 1+2. ציון אישי ובונוס תוצאה, לכל שחקן ---
  const players: PlayerScore[] = xi.map((slot) =>
    scorePlayer(slot, perf.get(slot.playerId), out.get(slot.teamId), rules),
  );

  // --- 4. הקפטן הדובידי (משנה את השחקן הרלוונטי במקום) ---
  applyDubidCaptain(xi, players, perf, out, rules);

  // --- 3. שערי הרכב וירטואליים ---
  const bonusLines = virtualGoals(xi, out, rules);

  const allPlayerLines = players.flatMap((p) => p.lines);
  const byReason = (...codes: string[]) =>
    round2(sum(allPlayerLines.filter((l) => codes.includes(l.reason))));

  const captainPoints = byReason(
    Reason.CaptainMultiplier,
    Reason.CaptainImmunity,
  );
  const resultPoints = byReason(Reason.ResultBonus);
  const personalPoints = round2(
    sum(
      allPlayerLines.filter(
        (l) =>
          l.reason !== Reason.ResultBonus &&
          l.reason !== Reason.CaptainMultiplier &&
          l.reason !== Reason.CaptainImmunity &&
          l.reason !== Reason.CaptainTransferred,
      ),
    ),
  );
  const virtualPoints = round2(sum(bonusLines));

  return {
    lineupId: lineup.lineupId,
    gameweekId: lineup.gameweekId,
    rulesetVersion: rules.version,
    players,
    bonusLines,
    personalPoints,
    resultPoints,
    captainPoints,
    virtualPoints,
    totalPoints: round2(
      players.reduce((a, p) => a + p.subtotal, 0) + sum(bonusLines),
    ),
  };
}

/* =================================================================== */
/* 1 + 2 · ציון אישי ובונוס תוצאה                                       */
/* =================================================================== */

function scorePlayer(
  slot: LineupSlot,
  perf: PlayerPerformance | undefined,
  outcome: TeamOutcome | undefined,
  rules: RuleSet,
): PlayerScore {
  const p = rules.personal;
  const pos = slot.position;
  const lines: ScoreLine[] = [];
  const add = (
    reason: ScoreLine['reason'],
    points: number,
    count?: number,
    meta?: Record<string, unknown>,
  ) => {
    if (points !== 0) lines.push({ reason, points: round2(points), count, playerId: slot.playerId, meta });
  };

  if (perf && perf.played) {
    if (perf.minutes > 0) {
      add(Reason.Minutes, perf.minutes >= 60 ? p.minutes60Plus : p.minutesPlayed, perf.minutes);
    }
    if (perf.goals) add(Reason.Goal, p.goal[pos] * perf.goals, perf.goals, { perGoal: p.goal[pos] });
    if (perf.assists) add(Reason.Assist, p.assist * perf.assists, perf.assists);
    if (perf.cleanSheet && perf.minutes >= p.cleanSheetMinMinutes) {
      add(Reason.CleanSheet, p.cleanSheet[pos]);
    }
    if (perf.goalsConceded >= p.goalsConcededPer) {
      add(
        Reason.GoalsConceded,
        p.goalsConceded[pos] * Math.floor(perf.goalsConceded / p.goalsConcededPer),
        perf.goalsConceded,
      );
    }
    if (perf.saves >= p.savesPer) {
      add(Reason.Saves, p.savesPoints * Math.floor(perf.saves / p.savesPer), perf.saves);
    }
    if (perf.penaltiesSaved) add(Reason.PenaltySaved, p.penaltySaved * perf.penaltiesSaved, perf.penaltiesSaved);
    if (perf.penaltiesMissed) add(Reason.PenaltyMissed, p.penaltyMissed * perf.penaltiesMissed, perf.penaltiesMissed);
    if (perf.ownGoals) add(Reason.OwnGoal, p.ownGoal * perf.ownGoals, perf.ownGoals);
    if (perf.yellowCards) add(Reason.Yellow, p.yellowCard * perf.yellowCards, perf.yellowCards);
    if (perf.redCards) add(Reason.Red, p.redCard * perf.redCards, perf.redCards);
  }

  // בונוס תוצאה — שייך לשחקן כי הוא נובע מהקבוצה שהוא מייצג בהרכב
  if (outcome) {
    const rb = rules.resultBonus;
    const eligible = rb.requireMinutes === 0 || (perf?.minutes ?? 0) >= rb.requireMinutes;
    const pts = rb[outcome.result];
    if (eligible && pts !== 0) {
      lines.push({
        reason: Reason.ResultBonus,
        points: pts,
        playerId: slot.playerId,
        teamId: outcome.teamId,
        meta: { result: outcome.result },
      });
    }
  }

  return {
    playerId: slot.playerId,
    teamId: slot.teamId,
    position: pos,
    isCaptain: !!slot.isCaptain,
    lines,
    subtotal: round2(sum(lines)),
  };
}

/* =================================================================== */
/* 4 · ★ הקפטן הדובידי                                                  */
/* =================================================================== */

function applyDubidCaptain(
  xi: LineupSlot[],
  players: PlayerScore[],
  perf: Map<string, PlayerPerformance>,
  outcomes: Map<string, TeamOutcome>,
  rules: RuleSet,
): void {
  const c = rules.captain;
  const captainSlot = xi.find((s) => s.isCaptain);
  if (!captainSlot) return;

  // -- שלב א׳: קפטן שלא ירד למגרש -> הכפולה עוברת לסגן --------------
  let effectiveSlot = captainSlot;
  let transferred = false;
  const captainMinutes = perf.get(captainSlot.playerId)?.minutes ?? 0;

  if (captainMinutes < c.transferToViceIfMinutesBelow) {
    const vice = xi.find((s) => s.isVice && s.playerId !== captainSlot.playerId);
    const viceMinutes = vice ? (perf.get(vice.playerId)?.minutes ?? 0) : 0;
    if (vice && viceMinutes >= c.transferToViceIfMinutesBelow) {
      effectiveSlot = vice;
      transferred = true;
    } else {
      return; // אין קפטן פעיל ואין סגן פעיל — אין מכפיל. תיעוד בלבד.
    }
  }

  const target = players.find((p) => p.playerId === effectiveSlot.playerId);
  if (!target) return;

  if (transferred) {
    target.lines.push({
      reason: Reason.CaptainTransferred,
      points: 0,
      playerId: target.playerId,
      meta: { from: captainSlot.playerId, reason: 'captain_did_not_play' },
    });
    target.isCaptain = true;
  }

  // -- שלב ב׳: חסינות כרטיסים ---------------------------------------
  const result = outcomes.get(target.teamId)?.result;
  const minutes = perf.get(target.playerId)?.minutes ?? 0;
  const immunityEarned =
    minutes >= c.immunity.minMinutes &&
    (result === undefined || c.immunity.requireResultIn.includes(result));

  if (immunityEarned) {
    const forgiven: string[] = [];
    if (c.immunity.yellowCards) forgiven.push(Reason.Yellow);
    if (c.immunity.redCards) forgiven.push(Reason.Red);

    const penalty = sum(target.lines.filter((l) => forgiven.includes(l.reason)));
    if (penalty < 0) {
      // הקו המקורי נשאר בפירוט (המשתמש רוצה לראות שהיה כרטיס),
      // והזיכוי מופיע לידו כשורה נפרדת. שקיפות > קיצור.
      target.lines.push({
        reason: Reason.CaptainImmunity,
        points: round2(-penalty),
        playerId: target.playerId,
        meta: { forgiven, minutes, result },
      });
    }
  }

  // -- שלב ג׳: המכפיל -----------------------------------------------
  const multiplied = target.lines.filter((l) =>
    c.appliesTo === 'personalAndResult'
      ? l.reason !== Reason.CaptainMultiplier && l.reason !== Reason.CaptainTransferred
      : l.reason !== Reason.ResultBonus &&
        l.reason !== Reason.CaptainMultiplier &&
        l.reason !== Reason.CaptainTransferred,
  );
  const base = round2(sum(multiplied));

  if (base > 0 || (c.applyToNegative && base !== 0)) {
    const extra = round2(base * (c.multiplier - 1));
    if (extra !== 0) {
      target.lines.push({
        reason: Reason.CaptainMultiplier,
        points: extra,
        playerId: target.playerId,
        meta: { multiplier: c.multiplier, base, appliesTo: c.appliesTo },
      });
    }
  }

  target.subtotal = round2(sum(target.lines));
}

/* =================================================================== */
/* 3 · שערי הרכב וירטואליים                                             */
/* =================================================================== */

function virtualGoals(
  xi: LineupSlot[],
  outcomes: Map<string, TeamOutcome>,
  rules: RuleSet,
): ScoreLine[] {
  const v = rules.virtualGoal;
  if (v.goalsPerVirtual <= 0) return [];

  // קבוצות ייחודיות: גם אם ליגה עתידית תאפשר 2 מאותה קבוצה,
  // השערים שלה לא ייספרו פעמיים.
  const teamGoals = new Map<string, number>();
  for (const slot of xi) {
    const o = outcomes.get(slot.teamId);
    if (o) teamGoals.set(slot.teamId, o.goalsFor);
  }
  if (teamGoals.size === 0) return [];

  if (v.aggregation === 'perTeam') {
    const lines: ScoreLine[] = [];
    for (const [teamId, goals] of [...teamGoals].sort()) {
      const count = Math.floor(goals / v.goalsPerVirtual);
      if (count > 0) {
        lines.push({
          reason: Reason.VirtualGoal,
          points: round2(v.points * count),
          count,
          teamId,
          meta: { teamGoals: goals },
        });
      }
    }
    return lines;
  }

  const total = [...teamGoals.values()].reduce((a, b) => a + b, 0);
  const count = Math.floor(total / v.goalsPerVirtual);
  if (count === 0) return [];
  return [
    {
      reason: Reason.VirtualGoal,
      points: round2(v.points * count),
      count,
      meta: {
        teamGoalsTotal: total,
        remainder: total % v.goalsPerVirtual,
        perTeam: Object.fromEntries(teamGoals),
      },
    },
  ];
}

/* =================================================================== */
/* מצרפים ועזרים                                                        */
/* =================================================================== */

/** דירוג עם טיפול בשוויון: אותו ניקוד = אותו מקום. */
export function rankGameweek(scores: LineupScore[]): Array<{ rank: number; score: LineupScore }> {
  const ordered = [...scores].sort((a, b) => b.totalPoints - a.totalPoints);
  const out: Array<{ rank: number; score: LineupScore }> = [];
  let lastPoints: number | null = null;
  let lastRank = 0;
  ordered.forEach((score, i) => {
    if (score.totalPoints !== lastPoints) {
      lastRank = i + 1;
      lastPoints = score.totalPoints;
    }
    out.push({ rank: lastRank, score });
  });
  return out;
}

/** פירוט לשמירה ב-DB ולתצוגה ב-UI. */
export function toBreakdown(score: LineupScore) {
  return {
    total: score.totalPoints,
    personal: score.personalPoints,
    result: score.resultPoints,
    captain: score.captainPoints,
    virtual: score.virtualPoints,
    rulesetVersion: score.rulesetVersion,
    players: score.players.map((p) => ({
      playerId: p.playerId,
      teamId: p.teamId,
      position: p.position,
      isCaptain: p.isCaptain,
      subtotal: p.subtotal,
      lines: p.lines.map((l) => ({ reason: l.reason, points: l.points, count: l.count, meta: l.meta })),
    })),
    bonuses: score.bonusLines,
  };
}

function toMap<T>(input: Map<string, T> | Record<string, T>): Map<string, T> {
  return input instanceof Map ? input : new Map(Object.entries(input));
}

/**
 * המרת שורות מה-DB לקלט המנוע.
 * שחקן ששיחק שני משחקים במחזור (השלמת משחק דחוי) — הנתונים נצברים.
 */
export function buildInputs(
  statRows: Array<Record<string, any>>,
  matchRows: Array<Record<string, any>>,
): { performances: Map<string, PlayerPerformance>; outcomes: Map<string, TeamOutcome> } {
  const acc = new Map<string, PlayerPerformance>();
  for (const r of statRows) {
    const cur =
      acc.get(r.player_id) ??
      ({
        playerId: r.player_id,
        teamId: r.team_id,
        position: r.position,
        minutes: 0, goals: 0, assists: 0, ownGoals: 0, yellowCards: 0, redCards: 0,
        saves: 0, penaltiesSaved: 0, penaltiesMissed: 0, goalsConceded: 0,
        cleanSheet: false, played: false,
      } as PlayerPerformance);
    cur.minutes += r.minutes ?? 0;
    cur.goals += r.goals ?? 0;
    cur.assists += r.assists ?? 0;
    cur.ownGoals += r.own_goals ?? 0;
    cur.yellowCards += r.yellow_cards ?? 0;
    cur.redCards += r.red_cards ?? 0;
    cur.saves += r.saves ?? 0;
    cur.penaltiesSaved += r.penalties_saved ?? 0;
    cur.penaltiesMissed += r.penalties_missed ?? 0;
    cur.goalsConceded += r.goals_conceded ?? 0;
    cur.played = cur.minutes > 0;
    cur.cleanSheet = cur.goalsConceded === 0 && cur.minutes > 0;
    acc.set(r.player_id, cur);
  }

  const outcomes = new Map<string, TeamOutcome>();
  const rank = { W: 2, D: 1, L: 0 } as const;
  for (const r of matchRows) {
    const prev = outcomes.get(r.team_id);
    if (!prev) {
      outcomes.set(r.team_id, {
        teamId: r.team_id,
        result: r.result,
        goalsFor: r.goals_for ?? 0,
        goalsAgainst: r.goals_against ?? 0,
      });
    } else {
      outcomes.set(r.team_id, {
        teamId: r.team_id,
        result: rank[r.result as 'W'] > rank[prev.result] ? r.result : prev.result,
        goalsFor: prev.goalsFor + (r.goals_for ?? 0),
        goalsAgainst: prev.goalsAgainst + (r.goals_against ?? 0),
      });
    }
  }

  return { performances: acc, outcomes };
}
